### PR TITLE
Stricter react linting

### PR DESCRIPTION
**Pull request by chrislloyd**
To test out locally:

```
# git clone this branch
> npm link

# cd into pinboard/webapp/app
> npm link eslint-config-pinterest
> eslint common/react
```

Comment here if you think rules should change / are too strict. Our code lints :+1: I'll merge this PR and bump version.

### CODE REVIEW

File: rules/react.js [hunk 1]
```diff
@@ -6,8 +6,156 @@ module.exports = {
...
+        'react/react-in-jsx-scope': 2,
+        // Restrict file extensions that may be required
+        // https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/require-extension.md
+        'react/require-extension': [2, { 'extensions': ['.js'] }],
+        // Prevent extra closing tags for components without children
+        // https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/self-closing-comp.md
+        'react/self-closing-comp': 2,
+        // Enforce spaces before the closing bracket of self-closing JSX elements
+        // https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-space-before-closing.md
+        'react/jsx-space-before-closing': [2, 'always'],
```
Comment (zackargyle) [inline rules/react.js]: This is in here twice.

File: rules/react.js [hunk 2]
```diff
@@ -6,8 +6,156 @@ module.exports = {
...
+        // Enforce component methods order
+        // https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/sort-comp.md
+        'react/sort-comp': [2, {
+            'order': [
+                'static-methods',
+                'lifecycle',
+                'rendering',
```
Comment (zackargyle) [inline rules/react.js]: I prefer rendering functions at the end. Easier to search out.

Comment (adamburmister) [inline rules/react.js]: I agree. Fitt's law: easier to scroll to bottom to find `render`.

Comment (chrislloyd) [inline rules/react.js]: Ha, my preference was justified with Fitt's law too — no scrolling is better than scrolling. Also, the render method tends to give you a very good overview of the structure of the file. Happy to change if you guys feel strongly about this.

Comment (KevinGrandon) [inline rules/react.js]: I'm used to having rendering at the end, so would prefer to go with that, but wouldn't bother me too much either way ¯\_(ツ)_/¯

File: rules/react.js [hunk 3]
```diff
@@ -6,8 +6,153 @@ module.exports = {
...
     rules: {
-        'no-unreferenced-globals': 0,
-        'react/jsx-uses-react': 2,
-        'react/jsx-uses-vars': 2
+        // Prevent missing displayName in a React component definition
+        // https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/display-name.md
+        'react/display-name': [0, { 'ignoreTranspilerName': false }],
+        // Forbid certain propTypes (any, array, object)
+        // https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/forbid-prop-types.md
+        'react/forbid-prop-types': [2, { 'forbid': ['any', 'array', 'object'] }],
```
Comment (KevinGrandon) [inline rules/react.js]: I'm finding this rule very difficult to work with given the current structure of our objects coming from the data layer. It's going to be quite difficult to write a validation for `context` and `user` props that we pass everywhere as well.

I recommend removing this rule from the PR.

**Comment by KevinGrandon** [comment]
We landed most of these rules into master, albeit commented out, and are currently working through enabling the rest of them.

**Pull request closed without merging**